Fix incorrect validation state when connecting many blocks

ptschip opened the pull request:
If we try to connect an invalid block and then subsequently try
to connect a valid one in the same ActivateBestChainStep() loop
then we end up with an incorrect validation state which then causes
us to set out pindexBestHeader to the wrong value , at the point
when we are wrapping up the parallel validation thread.  What we
need to do is ensure that each time we loop through and attempt
to connect a block , that we have a fresh validation state.

Futhermore in ActivateBestChain() where we "remember" the invalid
state that was returned in the bool "result" , we have to reset
that to true if any subsequent block is sucessfully connected
otherwise we will also potentially return an unsucessful result
when in fact we were finally successful.

ptschip committed:
Fix incorrect validation state when connecting many blocks

If we try to connect an invalid block and then subsequently try
to connect a valid one in the same ActivateBestChainStep() loop
then we end up with an incorrect validation state which then causes
us to set out pindexBestHeader to the wrong value , at the point
when we are wrapping up the parallel validation thread.  What we
need to do is ensure that each time we loop through and attempt
to connect a block , that we have a fresh validation state.

Furthermore in ActivateBestChain() where we "remember" the invalid
state that was returned in the bool "result" , we have to reset
that to true if any subsequent block is sucessfully connected
otherwise we will also potentially return an unsucessful result
when in fact we were finally successful.

sickpig commented:
@ptschip 	
code changes sounds good to me, you have a formatting issue thou: https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/553452003

ptschip committed:
After we get an invalid block then set the pindexBestHeader

   to be the most work chain.

Prior we were sometimes setting it to the chain tip at times and
also setting it after ProcessNewBlock() wheras we should be setting
it right after getting an invalid block.

ptschip committed:
Add another syncronization point to ctor.py

ptschip committed:
Move the sending of the block reject message to ActivateBestChain

The reject message was getting sent at the very end of the process
of connecting blocks, however, the rejected block may not necessarily
be the last block and therefore we could end up sending a reject for
the wrong block. Instead we send out the reject as soon as we find
the block is invalid and then keep iterating to connect the valid
chain.

gandrewstone commented on src/validation/validation.cpp:
```diff
@@ -3186,8 +3208,6 @@ bool ActivateBestChainStep(CValidationState &state,
             }
         }
-        if (fInvalidFound)
-            break; // stop processing more blocks if the last one was invalid.
```
why remove?  I think at the very least, the next line (notify the UI) should not be run if an invalid block is found (since there's no new tip)... subsequent lines also seem unnecesssary.  Is there a reason to continue processing blocks beyond an invalid one?

ptschip replied:
fixed

ptschip committed:
Break from loop when invalid block is found
